crates/ty_python_semantic/src/types/diagnostic.rs

```diff
@@ -1816,16 +1816,57 @@ pub(super) fn report_invalid_assignment(
     target_ty: Type,
     source_ty: Type,
 ) {
-    report_invalid_assignment_with_message(
-        context,
-        node,
-        target_ty,
-        format_args!(
-            "Object of type `{}` is not assignable to `{}`",
-            source_ty.display(context.db()),
-            target_ty.display(context.db()),
-        ),
-    );
+    // Handles the situation where the report naming is confusing, such as class with the same Name,
+    // but from different scopes.
+    let use_qualified = if let Some(target_class) = type_to_class_literal(target_ty, context.db()) {
+        if let Some(source_class) = type_to_class_literal(source_ty, context.db()) {
+            target_class != source_class
+                && target_class.name(context.db()) == source_class.name(context.db())
+        } else {
+            false
+        }
+    } else {
+        false
+    };
+
+    if use_qualified {
+        report_invalid_assignment_with_message(
+            context,
+            node,
+            target_ty,
+            format_args!(
+                "Object of type `{}` is not assignable to `{}`",
+                source_ty.qualified_display(context.db()),
+                target_ty.qualified_display(context.db()),
+            ),
+        );
+    } else {
+        report_invalid_assignment_with_message(
+            context,
+            node,
+            target_ty,
+            format_args!(
+                "Object of type `{}` is not assignable to `{}`",
+                source_ty.display(context.db()),
+                target_ty.display(context.db())
+            ),
+        );
+    }
+}
+
+fn type_to_class_literal<'db>(
+    ty: Type<'db>,
+    db: &'db dyn crate::Db,
+) -> Option<crate::types::class::ClassLiteral<'db>> {
+    match ty {
+        Type::ClassLiteral(class) => Some(class),
+        Type::NominalInstance(instance) => match instance.class {
+            crate::types::class::ClassType::NonGeneric(class) => Some(class),
+            crate::types::class::ClassType::Generic(alias) => Some(alias.origin(db)),
+        },
+        Type::EnumLiteral(enum_literal) => Some(enum_literal.enum_class(db)),
+        _ => None,
```

Comment:
There are a few more variants here where the class can pop up in the display:

```suggestion
        Type::NominalInstance(instance) => type_to_class_literal(Type::from(instance.class)),
        Type::EnumLiteral(enum_literal) => Some(enum_literal.enum_class(db)),
        Type::GenericAlias(alias) => Some(alias.origin(db)),
        Type::ProtocolInstance(ProtocolInstanceType { inner: Protocol::FromClass(class) }) => Some(class),
        Type::TypedDictType(typed_dict) => Some(typed_dict.defining_class(db)),
        Type::SubclassOf(subclass_of) => type_to_class_literal(Type::from(subclass_of.subclass_of().into_class()?)),
        _ => None,
```